File: merge-intervals/merge-intervals.swift
class Solution {
    func merge(_ intervals: [[Int]]) -> [[Int]] {
        var intervalArr = [(Int, Int)]()
        
        for array in intervals{
            intervalArr.append( (array[0], array[1]))
        }
        
        intervalArr.sort{ $0.0 < $1.0 }
        
        var ans = [[Int]]()
        
        var interval = (intervalArr[0].0, intervalArr[0].1)
        
        for (first, second) in intervalArr{
            if interval.1 >= first{
                interval.1 = max(interval.1, second)
            } else{
                ans.append([interval.0, interval.1])
                interval = (first, second)
            }
            
        }
        ans.append([interval.0, interval.1])
        return ans
    }
}

File: merge-intervals/merge-intervals.cpp
  struct Meeting{
        int time;
        bool isStarted;
        
        Meeting(int time, bool isStarted){
            this->time = time;
            this->isStarted = isStarted;
        }
    };
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector< Meeting > meetings;
        int roomsNeeded = 0;
        for(auto &interval : intervals){
            Meeting startMeeting = Meeting(interval[0], true);
            Meeting endMeeting = Meeting(interval[1], false);
            meetings.push_back(startMeeting);
            meetings.push_back(endMeeting);
        }
        
        sort(meetings.begin(), meetings.end(), [](auto const &first, auto const &second){
            if(first.time < second.time){
                return true;
            }else if(first.time == second.time){
                 return first.isStarted > second.isStarted;
            }else return false;
        });
        
        int rooms = 0;
        int size = meetings.size();
        vector< vector<int> > ans;
        int start = 0;
        for(int index = 0; index < size; index ++){
            if(meetings[index].isStarted){
                rooms ++;
                if(rooms == 1){
                    start = meetings[index].time;
                }
                roomsNeeded = max(roomsNeeded, rooms);
            }else{
                rooms --;
                if(rooms == 0){
                    ans.push_back({start,  meetings[index].time});
                }
            }
        }
        
        return ans;
    }
};